feat: decommission feature for pools

harshavardhana opened the pull request:
## Description
feat: decommission feature for pools

## Motivation and Context
```
λ mc admin decommission start alias/ http://minio{1...2}/data{1...4}
```

```
λ mc admin decommission status alias/
┌─────┬─────────────────────────────────┬──────────────────────────────────┬────────┐
│ ID  │ Pools                           │ Capacity                         │ Status │
│ 1st │ http://minio{1...2}/data{1...4} │ 439 GiB (used) / 561 GiB (total) │ Active │
│ 2nd │ http://minio{3...4}/data{1...4} │ 329 GiB (used) / 421 GiB (total) │ Active │
└─────┴─────────────────────────────────┴──────────────────────────────────┴────────┘
```

```
λ mc admin decommission status alias/ http://minio{1...2}/data{1...4}
Progress: ===================> [1GiB/sec] [15%] [4TiB/50TiB]
Time Remaining: 4 hours (started 3 hours ago)
```

```
λ mc admin decommission status alias/ http://minio{1...2}/data{1...4}
ERROR: This pool is not scheduled for decommissioning currently.
```

```
λ mc admin decommission cancel alias/
┌─────┬─────────────────────────────────┬──────────────────────────────────┬──────────┐
│ ID  │ Pools                           │ Capacity                         │ Status   │
│ 1st │ http://minio{1...2}/data{1...4} │ 439 GiB (used) / 561 GiB (total) │ Draining │
└─────┴─────────────────────────────────┴──────────────────────────────────┴──────────┘
```

> NOTE: Canceled decommission will not make the pool active again, since we might have
> Potentially partial duplicate content on the other pools, to avoid this scenario be
> very sure to start decommissioning as a planned activity.

```
λ mc admin decommission cancel alias/ http://minio{1...2}/data{1...4}
┌─────┬─────────────────────────────────┬──────────────────────────────────┬────────────────────┐
│ ID  │ Pools                           │ Capacity                         │ Status             │
│ 1st │ http://minio{1...2}/data{1...4} │ 439 GiB (used) / 561 GiB (total) │ Draining(Canceled) │
└─────┴─────────────────────────────────┴──────────────────────────────────┴────────────────────┘
```

## How to test this PR?
Needs `mc` as well, however most functionality should work as is. 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [x] Unit tests added/updated

harshavardhana committed:
feat: decommission feature for pools

```
λ mc admin decommission start alias/ http://minio{1...2}/data{1...4}
```

```
λ mc admin decommission status alias/
┌─────┬─────────────────────────────────┬──────────────────────────────────┬────────┐
│ ID  │ Pools                           │ Capacity                         │ Status │
│ 1st │ http://minio{1...2}/data{1...4} │ 439 GiB (used) / 561 GiB (total) │ Active │
│ 2nd │ http://minio{3...4}/data{1...4} │ 329 GiB (used) / 421 GiB (total) │ Active │
└─────┴─────────────────────────────────┴──────────────────────────────────┴────────┘
```

```
λ mc admin decommission status alias/ http://minio{1...2}/data{1...4}
Progress: ===================> [1GiB/sec] [15%] [4TiB/50TiB]
Time Remaining: 4 hours (started 3 hours ago)
```

```
λ mc admin decommission status alias/ http://minio{1...2}/data{1...4}
ERROR: This pool is not scheduled for decommissioning currently.
```

```
λ mc admin decommission cancel alias/
┌─────┬─────────────────────────────────┬──────────────────────────────────┬──────────┐
│ ID  │ Pools                           │ Capacity                         │ Status   │
│ 1st │ http://minio{1...2}/data{1...4} │ 439 GiB (used) / 561 GiB (total) │ Draining │
└─────┴─────────────────────────────────┴──────────────────────────────────┴──────────┘
```

> NOTE: Canceled decommission will not make the pool active again, since we might have
> Potentially partial duplicate content on the other pools, to avoid this scenario be
> very sure to start decommissioning as a planned activity.

```
λ mc admin decommission cancel alias/ http://minio{1...2}/data{1...4}
┌─────┬─────────────────────────────────┬──────────────────────────────────┬────────────────────┐
│ ID  │ Pools                           │ Capacity                         │ Status             │
│ 1st │ http://minio{1...2}/data{1...4} │ 439 GiB (used) / 561 GiB (total) │ Draining(Canceled) │
└─────┴─────────────────────────────────┴──────────────────────────────────┴────────────────────┘
```

klauspost commented on cmd/admin-handlers-pools.go:
```diff
@@ -0,0 +1,179 @@
+
+	defer logger.AuditLog(ctx, w, r, mustGetClaimsFromToken(r))
+
+	objectAPI, _ := validateAdminReq(ctx, w, r, iampolicy.DecommissionAdminAction)
```
Maybe we should have less restrictions on Listing?

harshavardhana replied:
@klauspost like?

klauspost replied:
Maybe?
```suggestion
	objectAPI, _ := validateAdminReq(ctx, w, r, iampolicy.ServerInfoAdminAction)
```

harshavardhana replied:
Decommission is a separate granular policy action to allow for specific credentials to perform this operation.

Such as `admin:*` allows for this normally for consoleAdmin 

However they can come up with a credential that is specific.. to do the job and won't be needed for anything else.

harshavardhana replied:
But yes it's a question does customers need this granularity.

harshavardhana replied:
done allows two actions now.

klauspost commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,881 @@
+	if !found {
+		pd.QueuedBuckets = append(pd.QueuedBuckets, bucket)
+	}
+}
```
```suggestion
func (pd *PoolDecommissionInfo) bucketPush(bucket string) {
	for _, b := range pd.QueuedBuckets {
		if b == bucket {
			return
		}
	}
	pd.QueuedBuckets = append(pd.QueuedBuckets, bucket)
}
```
I can't tell if this should also check if the bucket has been decomissioned already?

klauspost commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,881 @@
+//msgp:decode ignore versionsSorter
+//msgp:marshal ignore versionsSorter
+//msgp:unmarshal ignore versionsSorter
+//msgp:size ignore versionsSorter
```
```suggestion
//msgp:ignore versionsSorter
```
You are ignoring everything, right?

klauspost commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,881 @@
+			versionsSorter(fivs.Versions).reverse()
+
+			for _, version := range fivs.Versions {
+				// TODO: Skip transitioned objects for now.
```
Shouldn't this be fixed before we can release?

harshavardhana replied:
No we decided to do this later @klauspost going to add a graceful to error when it finds transition buckets

klauspost commented on cmd/erasure-server-pool.go:
```diff
@@ -42,10 +42,15 @@ import (
+	serverPools   []*erasureSets
 
-	// Shut down async operations
+	// Shutdown async operations
```
```suggestion
	// Shut down async operations
```

> When you must choose between one or the other, think about that space between the words shut and down. Then imagine adding a word in that space: shutting something down. That might help you remember that the two-word option is a phrasal verb.

> Don’t let bad grammar shut down your ability to communicate the message!

:D

Source: https://www.patrickkphillips.com/grammar/shut-down-or-shutdown-should-it-be-one-word-or-two/

klauspost commented on cmd/object-api-datatypes.go:
```diff
@@ -46,6 +46,16 @@ const (
 type StorageInfo = madmin.StorageInfo
 
+// TotalUsableCapacity - total usable capacity
+func TotalUsableCapacity(s StorageInfo) float64 {
```
Can't we make this int64?

I don't think there is any reasonable chance it will overflow.

klauspost commented on docs/distributed/POOLS.md:
````diff
@@ -0,0 +1,52 @@
+ERROR: This pool is not scheduled for decommissioning currently.
+```
+
+### Canceling a decommission?
````
I don't really understand what "canceled" state means. What does it do? Is it a "pause" function?

harshavardhana replied:
It's more like stop the decommission for now.

klauspost replied:
Couldn't we name it "pause", if that is what we want to do?

Canceled is fine, if it just leaves it in the current state, but from the description is looks more like a "pause/resume" functionality.

harshavardhana replied:
Yeah well @klauspost I thought cancel is what happens more than pause and there is no resume CLI it's implicit.

You just have to start again what you did before, if necessary it will resume a previously canceled operation.

Just don't want to add anymore CLI. Let me add more details..

klauspost reviewed (commented):
Very solid!

Nothing major, just a few minor nits.

krisis commented on cmd/erasure-common.go:
```diff
@@ -22,6 +22,38 @@ import (
+			}
+
+			mu.Lock()
+			newDisks = append(newDisks, disks[i-1])
```
Does the order in which online disks appear in `newDisks` matter?

harshavardhana replied:
No it does not @krisis

krisis commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,881 @@
+func (p *poolMeta) DecommissionFailed(idx int) bool {
+	p.Pools[idx].LastUpdate = time.Now().UTC()
+	if p.Pools[idx].Decommission != nil {
+		p.Pools[idx].Decommission.StartTime = time.Time{}
```
why is `Decommission.StartTime` being reset on failure? Wouldn't it be useful to show start time in pool-decommission-status output?

harshavardhana replied:
Because we need to distinguish this with the difference to if the decommission is currently running or not.

krisis commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,881 @@
+	return p.Pools[idx].Decommission.bucketsToDecommission()
+}
+
+func (p *poolMeta) QueuedBuckets(idx int, buckets []BucketInfo) {
```
```suggestion
func (p *poolMeta) EnqueueBuckets(idx int, buckets []BucketInfo) {
```

harshavardhana replied:
It was supposed to be QueueBuckets() not Queued!

krisis commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,881 @@
+	binary.LittleEndian.PutUint16(data[0:2], poolMetaFormat)
+	binary.LittleEndian.PutUint16(data[2:4], poolMetaVersion)
+
+	buf, err := p.MarshalMsg(data)
```
Shouldn't this be,
```suggestion
	buf, err := p.MarshalMsg(data[4:])
```

harshavardhana replied:
No message pack returns the `data` first 4 bytes without interpretation @krisis - it uses the bytes that are not used automatically. 

The returned `buf` has all the bytes needed included the first 4 bytes.

harshavardhana replied:
check the implementation in `cmd/tier.go`

krisis commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,881 @@
+
+//go:generate msgp -file $GOFILE -unexported
+type poolMeta struct {
+	Version int          `msg:"v"`
```
Is `poolMeta.Version` needed in addition to `poolMetaFormat` and `poolMetaVersion` that are saved as part of `poolMeta` 'headers'?

harshavardhana replied:
yes, we need for further flexibility - the format is the format of this data structure. This version however allows for further format changes if needed.

krisis commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,881 @@
+func (p poolMeta) IsBucketDecommissioned(idx int, bucket string) bool {
+	if p.Pools[idx].Decommission == nil {
+		// Decommission not in progress, return 'true' to
+		// avoid any decommissioning.
```
This comment is not clear to me. Could you give an example situation where we may need to indicate that a given bucket is being decommissioned while `poolMeta` has no such information in it?

harshavardhana replied:
When you are resuming a decommission from a failed or canceled operation we need to restart from where it stopped last.

harshavardhana replied:
However this should return 'false' to decommission not the other way round.

harshavardhana replied:
Actually no - so this code is meant for a situation where we check if bucket is decommissioned only when decommission is in progress.

harshavardhana replied:
this code path should never be triggered otherwise.

krisis reviewed (commented):
Here are some initial comments.

krisis commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,909 @@
+}
+
+func (p *poolMeta) DecommissionComplete(idx int) bool {
+	p.Pools[idx].LastUpdate = time.Now().UTC()
```
Shouldn't we update `LastUpdate` only when `p.Pools[idx].Decommission` is non-nil?

harshavardhana replied:
This is called only when Decommission is `non-nil` @krisis

krisis commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,909 @@
+		}()
+		return meta.save(ctx, z.serverPools)
+	}
+	return nil
```
We could do something like this to avoid calling `DecommissionComplete` twice.
```suggestion
	meta := z.poolMeta
	if meta.DecommissionComplete(idx) {
	    if err = meta.save(ctx, z.serverPools); err == nil {
	        globalNotificationSys.ReloadPoolMeta(ctx)
	    }
	}
	return err
```

krisis commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,909 @@
+func (z *erasureServerPools) ReloadPoolMeta(ctx context.Context) (err error) {
+	meta := poolMeta{}
+
+	if _, err = meta.load(ctx, z.serverPools[0], z.serverPools); err != nil {
```
Would saving pool metadata on first pool work when it is the pool that is being decommissioned?

harshavardhana replied:
yes it will work @krisis - reads are allowed.

krisis commented on cmd/erasure-server-pool-decom.go:
```diff
@@ -0,0 +1,909 @@
+				// gr.Close() is ensured by decommissionObject().
+				if err = z.decommissionObject(ctx, bName, gr); err != nil {
+					logger.LogIf(ctx, err)
+					z.poolMeta.CountItem(idx, version.Size, true)
```
We should hold `z.poolMetaMutex` locks when calling `z.poolMeta.CountItem` too.

harshavardhana committed:
decommission missing metadata objects from .minio.sys/

additionally also error out if we find any transitioned
buckets and disallow decommissioning such deployments.

minio-trusted committed:
add a line for 'restart'ing decommission

klauspost reviewed (approved):
LGTM. Will of course require testing. And tiering, as you mention.
